libsolidity/ast/AsmJsonConverter.cpp

```diff
@@ -0,0 +1,272 @@
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @date 2019
+ * @author julius <djudju@protonmail.com>
+ * Converts inline assembly AST to JSON format
+ */
+
+#include <libsolidity/ast/AsmJsonConverter.h>
+#include <libyul/AsmData.h>
+//#include <libyul/AsmDataForward.h>
+#include <liblangutil/Exceptions.h>
+
+
+#include <libdevcore/CommonData.h>
+
+
+using namespace std;
+
+namespace dev
+{
+namespace solidity
+{
+
+Json::Value AsmJsonConverter::createAstNode(langutil::SourceLocation _location, string _nodeType) const
+{
+	Json::Value ret(Json::objectValue);
+	ret["nodeType"] = _nodeType;
+	int length = -1;
+	if (_location.start >= 0 && _location.end >= 0)
+		length = _location.end - _location.start;
+	ret["src"] = to_string(_location.start) + ":" + to_string(length) + ":" + to_string(m_sourceIndex);
+	return ret;
+}
+
+template <class T>
+Json::Value AsmJsonConverter::vectorOfVariantsToJson(vector<T> const& _vec ) const {
+	Json::Value ret(Json::arrayValue);
+	// TODO check what is returned when _vec is empty? nullPtr or []
+	for (auto const& var: _vec) {
+		ret.append(boost::apply_visitor(*this, var));
+	}
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::Block const& _node) const
+{
+	Json::Value ret = createAstNode(_node.location, "Block");
+	ret["statements"] = vectorOfVariantsToJson(_node.statements);
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::TypedName const& _node) const
+{
+	solAssert(!_node.name.empty(), "Invalid variable name.");
+	Json::Value ret = createAstNode(_node.location, "TypedName");
+	ret["name"] = _node.name.str();
+	ret["type"] = _node.type.str();
+	//NOTE isn't it weird that both are YulStrings but they are referenced differently in AsmData.h?
+	return ret;
+}
+
+Json::Value AsmJsonConverter::operator()(yul::Literal const& _node) const
+{
+	Json::Value ret = createAstNode(_node.location, "Literal");
+	switch (_node.kind)
+	{
+	case yul::LiteralKind::Number:
+		solAssert(isValidDecimal(_node.value.str()) || isValidHex(_node.value.str()), "Invalid number literal");
+		ret["type"] = _node.type.str(); //_node.value.str() + appendTypeName(_node.type);
+		break;
+	case yul::LiteralKind::Boolean:
+		// ?? what does this notation even mean >"true"_yulstring< ?
+//		solAssert(_node.value == "true"_yulstring || _node.value == "false"_yulstring, "Invalid bool literal.");
+		ret["type"] =  "Boolean"; //NOTE should it be like this? : ((_node.value == "true"_yulstring) ? "true" : "false") + appendTypeName(_node.type);
+		break;
+	case yul::LiteralKind::String:
+		ret["type"] = "String";
+		break;
+	default:
```

Comment:
Better to not use a default case, then the compiler will create an error if we add new items to the enum.

Reply:
It will not be possible to compile solc in that case.

Reply:
ah ok